java/squeek/appleskin/helpers/FoodHelper.java

```diff
@@ -40,13 +44,93 @@ public static boolean isRotten(ItemStack itemStack)
 		if (!isFood(itemStack))
 			return false;
 
-		for (Pair<StatusEffectInstance, Float> effect : itemStack.getItem().getFoodComponent().getStatusEffects())
-		{
-			if (effect.getFirst() != null && effect.getFirst().getEffectType() != null && effect.getFirst().getEffectType().getType() == StatusEffectType.HARMFUL)
-			{
+		for (Pair<StatusEffectInstance, Float> effect : itemStack.getItem().getFoodComponent().getStatusEffects()) {
+			if (effect.getFirst() != null && effect.getFirst().getEffectType() != null && effect.getFirst().getEffectType().getType() == StatusEffectType.HARMFUL) {
 				return true;
 			}
 		}
 		return false;
 	}
+
+	public static float getEstimatedHealthIncrement(ItemStack itemStack, PlayerEntity player)
+	{
+		if (!isFood(itemStack))
+			return 0;
+
+		if (!player.canFoodHeal()) {
+			return 0;
+		}
+
+		HungerManager stats = player.getHungerManager();
+		World world = player.getEntityWorld();
+
+		FoodValues foodValues = getModifiedFoodValues(itemStack, player);
+
+		int foodLevel = Math.min(stats.getFoodLevel() + foodValues.hunger, 20);
+		float healthIncrement = 0;
+
+		// health for natural regen
+		if (foodLevel >= 18.0F && world != null && world.getGameRules().getBoolean(GameRules.NATURAL_REGENERATION)) {
+			float saturationLevel = Math.min(stats.getSaturationLevel() + foodValues.getSaturationIncrement(), (float)foodLevel);
+			float exhaustionLevel = HungerHelper.getExhaustion(player);
+			healthIncrement = getEstimatedHealthIncrement(foodLevel, saturationLevel, exhaustionLevel);
+		}
+
+		// health for regeneration effect
+		for (Pair<StatusEffectInstance, Float> effect : itemStack.getItem().getFoodComponent().getStatusEffects()) {
+			StatusEffectInstance effectInstance = effect.getFirst();
+			if (effectInstance != null && effectInstance.getEffectType() == StatusEffects.REGENERATION) {
+				int amplifier = effectInstance.getAmplifier();
+				int duration = effectInstance.getDuration();
+				if (effectInstance.isPermanent()) {
+					duration = Integer.MAX_VALUE;
+				}
+				healthIncrement += (float)Math.floor(duration / (50 >> amplifier));
+				break;
+			}
+		}
+
+		return healthIncrement;
+	}
+
+	public static float getEstimatedHealthIncrement(int foodLevel, float saturationLevel, float exhaustionLevel)
+	{
+
+		float health = 0;
+		float exhaustionForRegen = 6.0F;
+		float exhaustionForConsumed = 4.0F;
+
+		// using timer is still used to avoid some boundary problems
+		int foodStarvationTimer = 0;
+		while (foodLevel >= 20 && saturationLevel > 0) {
+			if (exhaustionLevel > exhaustionForConsumed) {
```

Comment:
I think the timers are unecessary. If I understand correctly, the problem they are trying to solve is the 'double exhaustion' tick, but that can be done by doing:

```suggestion
		int foodStarvationTimer = 0;
		while (foodLevel >= 20 && saturationLevel > 0) {
			while (exhaustionLevel > exhaustionForConsumed) {
```
and getting rid of the foodStarvationTimer